ra/ra.go

```diff
@@ -1424,18 +1392,54 @@ func (ra *RegistrationAuthorityImpl) NewOrder(ctx context.Context, req *rapb.New
 		return nil, err
 	}
 
-	// TODO(#2955): Replace this with the batched methods
+	now := ra.clk.Now().UnixNano()
+	existingAuthz, err := ra.SA.GetAuthorizations(ctx, &sapb.GetAuthorizationsRequest{
+		RegistrationID: req.RegistrationID,
+		Now:            &now,
+		Domains:        parsedCSR.DNSNames,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	gotAuthzFor := make(map[string]bool, len(parsedCSR.DNSNames))
+	for _, v := range existingAuthz.Authz {
+		gotAuthzFor[*v.Domain] = true
+		order.Authorizations = append(order.Authorizations, *v.Authz.Id)
+	}
+
+	var newAuthz []*corepb.Authorization
 	for _, name := range parsedCSR.DNSNames {
-		authz, err := ra.NewAuthorization(ctx, core.Authorization{
-			Identifier: core.AcmeIdentifier{
-				Type:  core.IdentifierDNS,
-				Value: name,
-			},
-		}, *req.RegistrationID)
+		if _, present := gotAuthzFor[name]; present {
+			continue
+		}
+		identifier := core.AcmeIdentifier{Value: name, Type: core.IdentifierDNS}
+		if err := ra.PA.WillingToIssue(identifier); err != nil {
+			return nil, err
+		}
+		if err := ra.checkPendingAuthorizationLimit(ctx, *req.RegistrationID); err != nil {
+			return nil, err
+		}
+		if err := ra.checkInvalidAuthorizationLimit(ctx, *req.RegistrationID, identifier.Value); err != nil {
+			return nil, err
+		}
+		pa, err := ra.createPendingAuthz(ctx, *req.RegistrationID, identifier)
```

Comment:
Since this is a new method, and is only used to generate the protobuf form of an authz, let's return a protobuf directly rather than converting.

Reply:
It is also used in `NewAuthorization` to create the pending authorization that is passed to `sa.NewPendingAuthorization` in order to reduce code duplication.

Since `NewOrder` calls it `len(domains)` times and `NewAuthorization` only calls it once it probably still makes sense to return the protobuf and just convert it to a `core.Authorization` once in `NewAuthorization`.